p2p/types/request_map.go

```diff
@@ -0,0 +1,78 @@
+package types
+
+import sync "sync"
+
+type RequestMap struct {
+	sync.Mutex
+	maxcap   uint32
+	elements []*Request
+	nonces   uint32
+}
+
+func (rm *RequestMap) Get() *Request {
+	rm.Lock()
+	defer rm.Unlock()
+
+	rm.nonces++
+	nonce := rm.nonces
+	newreq := &Request{nonce: nonce, ch: make(chan Work)}
+	rm.elements = append(rm.elements, newreq)
+	return newreq
+}
+
+func (rm *RequestMap) Find(nonce uint32) (uint32, chan Work, bool) {
+	rm.Lock()
+	defer rm.Unlock()
+
+	var request *Request
+	var exists bool
+	var index int
+
+	for i := 0; i < len(rm.elements); i++ {
+		if rm.elements[i].nonce == nonce {
+			exists = true
+			index = i
+			request = rm.elements[i]
+		}
+	}
+
+	if exists {
+		rm.elements[index] = nil
+		fhalf := rm.elements[:index]
+		shalf := rm.elements[index+1:]
+		rm.elements = append(fhalf, shalf...)
+		return request.nonce, request.ch, exists
+	}
+
+	return nonce, nil, false
+}
+
+func (rm *RequestMap) Delete(nonce uint32) bool {
```

Comment:
Ditto my comments above for Golang style:

1. Use range for the loop
2. Max index an int pointer, and use its existence as the value for `exists`
3. Make the append of the first and second half a single statement